renderdoc/driver/ihv/nv/nv_counters.cpp

```diff
@@ -0,0 +1,380 @@
+
+  NvPmApi *mNvPmApi;
+
+  uint32_t mCurrentCounterId;
+};
+
+static bool NvPmResultFails(NVPMRESULT actual, char const *failMsg)
+{
+  if(actual != NVPM_OK)
+  {
+    RDCWARN("NV GPU performance counters could not %s (code = %u)\n", failMsg, actual);
+    return true;
+  }
+  return false;
+}
+
+int NvPmCountCounters(NVPMCounterID unCounterID, const char *pcCounterName, void *pUserData)
+{
+  uint32_t *pNumCounters = static_cast<uint32_t *>(pUserData);
+  *pNumCounters += 1;
+  return NVPM_OK;
+}
+
+int NvPmGatherCounters(NVPMCounterID unCounterID, const char *pcCounterName, void *pUserData)
+{
+  EnumCountersCtx *pEnumCtx = static_cast<EnumCountersCtx *>(pUserData);
+
+  const uint32_t i = pEnumCtx->mCurrentCounterId;
+
+  GPUCounter globalId = static_cast<GPUCounter>(static_cast<uint32_t>(GPUCounter::FirstNvidia) + i);
+  pEnumCtx->mExternalIds[i] = globalId;
+  pEnumCtx->mInternalIds[i] = unCounterID;
+
+  CounterDescription &desc = pEnumCtx->mExternalDescriptors[i];
+
+  NVPMUINT64 Attribute = 0;
+
+  pEnumCtx->mNvPmApi->GetCounterAttribute(unCounterID, NVPMA_COUNTER_TYPE, &Attribute);
+  NVPMCOUNTERTYPE Type = static_cast<NVPMCOUNTERTYPE>(Attribute);
+
+  switch(Type)
+  {
+    case NVPM_CT_GPU: desc.category = "GPU"; break;
+
+    case NVPM_CT_OGL: desc.category = "D3D"; break;
+
+    case NVPM_CT_D3D: desc.category = "D3D"; break;
+
+    case NVPM_CT_SIMEXP: desc.category = "SIMEXP"; break;
+
+    case NVPM_CT_AGGREGATE: desc.category = "AGGREGATE"; break;
+
+    case NVPM_CT_USER: desc.category = "USER"; break;
+  };
+
+  pEnumCtx->mNvPmApi->GetCounterAttribute(unCounterID, NVPMA_COUNTER_DISPLAY, &Attribute);
+  NVPMCOUNTERDISPLAY DisplayType = static_cast<NVPMCOUNTERDISPLAY>(Attribute);
+
+  pEnumCtx->mNvPmApi->GetCounterAttribute(unCounterID, NVPMA_COUNTER_DOMAIN, &Attribute);
+
+  pEnumCtx->mNvPmApi->GetCounterAttribute(unCounterID, NVPMA_COUNTER_VALUE_TYPE, &Attribute);
+  NVPMCOUNTERVALUETYPE ValueType = static_cast<NVPMCOUNTERVALUETYPE>(Attribute);
+
+  pEnumCtx->mInternalDescriptors[i] = (DisplayType << 1) | ValueType;
+
+  if(ValueType == NVPM_VALUE_TYPE_UINT64)
+  {
+    if(DisplayType == NVPM_CD_RATIO)
+    {
+      desc.unit = CounterUnit::Ratio;
+      desc.resultType = CompType::Double;
+      desc.resultByteWidth = sizeof(double);
+    }
+    else
+    {
+      desc.unit = CounterUnit::Absolute;
+      desc.resultType = CompType::UInt;
+      desc.resultByteWidth = sizeof(uint64_t);
+    }
+  }
+  else
+  {
+    if(DisplayType == NVPM_CD_RATIO)
+    {
+      RDCWARN(
+          " normalization for counters with DisplayType == NVPM_CD_RATIO and ValueType == "
+          "NVPM_VALUE_TYPE_FLOAT64 is unhandled");
+    }
+    desc.unit = CounterUnit::Ratio;
+    desc.resultType = CompType::Double;
+    desc.resultByteWidth = sizeof(double);
+  }
+
+  char Description[512];
+  NVPMUINT DescriptionSize = sizeof(Description);
+  pEnumCtx->mNvPmApi->GetCounterDescription(unCounterID, Description, &DescriptionSize);
```

Comment:
Is it possible to call this function twice, to first get the size and then get the actual text? In case the description is longer than 512 characters.